### PR TITLE
fix: integration tests for OrderBook market prices

### PR DESCRIPTION
**Pull request by wischli**
# Description

* Fixes integration tests for #1674
* Adds shortcut to orderbook for same currencies to not require market feed
* Adds integration test for edge case: `Increase -> Fulfill -> Decrease -> Fulfill partially -> Increase more than pending`

## TODO

- [x] Discuss and fix `increase_fulfill_decrease_fulfill_partial_increase` 

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: pallets/order-book/src/lib.rs [hunk 1]
```diff
@@ -868,6 +868,10 @@ pub mod pallet {
 			currency_out: Self::CurrencyId,
 			amount_out: Self::Balance,
 		) -> Result<Self::Balance, DispatchError> {
+			if currency_in == currency_out {
+				return Ok(amount_out);
+			}
+
```
Comment (lemunozm) [inline pallets/order-book/src/lib.rs:874]: I fully agree with this addition. Thanks!

File: runtime/integration-tests/src/generic/cases/liquidity_pools.rs [hunk 2]
```diff
@@ -3889,48 +3865,65 @@ mod development {
...
 						pool_id,
 						invest_amount_foreign_denominated,
 						investor.clone(),
 						foreign_currency,
 					);
+					let swap_order_id = pallet_foreign_investments::Swaps::<T>::swap_id_from(
+						&investor,
+						default_investment_id::<T>(),
+						pallet_foreign_investments::Action::Investment,
+					)
```
Comment (lemunozm) [inline runtime/integration-tests/src/generic/cases/liquidity_pools.rs:3903]: Not sure if it could make sense to expose this under a `RuntimeAPI`. How do Apps know which order placed by FI needs to be fulfilled?

@cc @mustermeiszer

File: runtime/integration-tests/src/generic/cases/liquidity_pools.rs [hunk 3]
```diff
@@ -3942,9 +3935,24 @@ mod development {
 						DEFAULT_DOMAIN_ADDRESS_MOONBEAM,
 						decrease_msg_pool_swap_amount
 					));
+					assert!(frame_system::Pallet::<T>::events().iter().any(|e| {
```
Comment (lemunozm) [inline runtime/integration-tests/src/generic/cases/liquidity_pools.rs:3938]: Super NIT. There is a `find_event()` in `generic/utils/mod.rs`

**Review by lemunozm — COMMENTED** [review]
Thanks for these update! I like what I see. 

I'll check wrong use case. 🙌🏻

File: runtime/integration-tests/src/generic/cases/liquidity_pools.rs [hunk 4]
```diff
@@ -4102,8 +4137,132 @@ mod development {
...
+					let swap_order_id = pallet_foreign_investments::Swaps::<T>::swap_id_from(
+						&investor,
+						default_investment_id::<T>(),
+						pallet_foreign_investments::Action::Investment,
+					)
+					.expect("Swap order exists; qed");
+					assert_ok!(pallet_order_book::Pallet::<T>::fill_order(
+						RawOrigin::Signed(trader.clone()).into(),
+						swap_order_id,
+						invest_amount_pool_denominated / 4 * 3
```
Comment (lemunozm) [inline runtime/integration-tests/src/generic/cases/liquidity_pools.rs:4218]: I think this is wrong. The multiplication has priority. So it is: `AMOUNT / 12`

Comment (wischli) [reply runtime/integration-tests/src/generic/cases/liquidity_pools.rs:4218]: It should not, rust math is distributive such that `x * y  / z = (x * y) / z`. Safety first, I checked by doing `3 * invest_amount / 4` and it didn't change anything.

Comment (wischli) [reply runtime/integration-tests/src/generic/cases/liquidity_pools.rs]: I think you are mixing it up with the order of operations for addition/subtraction and multiplication/division: `a + b + c = a + (b * c)`

File: runtime/integration-tests/src/generic/cases/liquidity_pools.rs (hunk 2 above)
Comment (wischli) [reply runtime/integration-tests/src/generic/cases/liquidity_pools.rs:3903]: I think it's fine to query `ForeignIdToSwapId::<T>::get((account, investment_id, action))` for Apps which would be `api.query.foreignInvestments.foreignIdToSwapId([account, investmentId, action['Investment'])`.

File: runtime/integration-tests/src/generic/cases/liquidity_pools.rs (hunk 3 above)
Comment (wischli) [reply runtime/integration-tests/src/generic/cases/liquidity_pools.rs:3938]: I don't find this API to be more helpful here because all foreign investment integration tests operate in a single closure of `env` such that for multiple of these events it is better to directly look for the exact same one as we are asserting at least two values here, not only a single one.

But I will definitely remember this for the next integration tests.

File: runtime/integration-tests/src/generic/cases/liquidity_pools.rs (hunk 2 above)
Comment (lemunozm) [reply runtime/integration-tests/src/generic/cases/liquidity_pools.rs:3903]: Make sense!

File: runtime/integration-tests/src/generic/cases/liquidity_pools.rs (hunk 4 above)
Comment (lemunozm) [reply runtime/integration-tests/src/generic/cases/liquidity_pools.rs]: Yeah, in the same way `*` takes more priority over `+`, I thought it was for  `*` and `/`. Thanks for enlighten me 🙌🏻

**Review by lemunozm — APPROVED** [review]
Approved! Amazing work with integration-tests and debugging @wischli!

Feel free to merge it. I've also added a UT for the swap ids case: [5536efe](https://github.com/centrifuge/centrifuge-chain/pull/1674/commits/5536efe134efad923d01f116a9554468f919dca6)

Right now, it fails (as expected) and your commit should fix it.

File: runtime/integration-tests/src/generic/cases/liquidity_pools.rs (hunk 3 above)
Comment (wischli) [reply runtime/integration-tests/src/generic/cases/liquidity_pools.rs:3938]: Thanks for the explanation. Next time, I touch the integration tests, I will re-check whether it makes sense to switch to that API :)
